mpi4jax/_src/tokenizer.py

```diff
@@ -0,0 +1,154 @@
+        new_branch_fn = lambda token, *args: _token_forwarding(
+            jax.core.jaxpr_as_fun(branch), token
+        )(*args)
+        new_jaxpr = jax.make_jaxpr(new_branch_fn)(token, *safe_map(read, other_vars))
+        branch_jaxprs.append(new_jaxpr)
+    bind_params["branches"] = tuple(branch_jaxprs)
+    bind_params["linear"] = (False,) + bind_params["linear"]
+    ans = eqn.primitive.bind(
+        read(cond_var), token, *safe_map(read, other_vars), **bind_params
+    )
+    token = ans[0]
+    ans = ans[1:]  # Drop the token.
+    return token, ans
+
+
+def _override_tokens(jaxpr, consts, token, *args):
+    if token is None:  # Create a new token if one is not passed.
+        token = jax.lax.create_token()
+
+    def read(v):
+        if type(v) is jax.core.Literal:
+            return v.val
+        else:
+            return env[v]
+
+    def write(v, val):
+        env[v] = val
+
+    env = {}
+    write(jax.core.unitvar, jax.core.unit)
+    safe_map(write, jaxpr.constvars, consts)
+    safe_map(write, jaxpr.invars, args)
+    for eqn in jaxpr.eqns:
+        # Here, we override the original primatives with our own
+        # token forcing method.
+        if eqn.primitive in token_override_registry:
+            token_override = token_override_registry[eqn.primitive]
+            ans = token_override(
+                safe_map(read, eqn.invars), new_token=token, **eqn.params
+            )
+            # We pass along the new token returned by the above binding
+            token = ans[-1]
+            safe_map(write, eqn.outvars, ans)
+        else:
+            if eqn.primitive is xla.xla_call_p:
+                token, ans = xla_call_overrride(read, eqn, token)
+            elif eqn.primitive is jax.lax.scan_p:
+                token, ans = scan_override(read, eqn, token)
+            elif eqn.primitive is jax.lax.while_p:
+                token, ans = while_override(read, eqn, token)
+            elif eqn.primitive is jax.lax.cond_p:
+                token, ans = cond_override(read, eqn, token)
+            # Here, we are just reapplying the original operation if
+            # not part of our communication protocol.
+            # This code is mostly taken form jax.core.eval_jaxpr
+            else:
+                subfuns, bind_params = eqn.primitive.get_bind_params(eqn.params)
+                ans = eqn.primitive.bind(
+                    *subfuns, *safe_map(read, eqn.invars), **bind_params
+                )
+            if eqn.primitive.multiple_results:
+                safe_map(write, eqn.outvars, ans)
+            else:
+                write(eqn.outvars[0], ans)
+    return (token,) + tuple(safe_map(read, jaxpr.outvars))
+
+
+def _token_forwarding(f, token=None):
+    def wrapper(*args, **kwargs):
+        jaxpr = jax.make_jaxpr(f)(*args, **kwargs)
+        return _override_tokens(jaxpr.jaxpr, jaxpr.consts, token, *args, **kwargs)
+
+    return wrapper
+
+
+def auto_tokenize(f, token=None):
```

Comment:
It is not clear to me how the `token` argument can be used here. Tokens are usually tracers, so I don't see how it makes sense to pass these to a function transformation that may be applied only once (e.g. at module import time).

If JAX guarantees in-order execution outside of JIT then we wouldn't even need this argument, since you can just do `auto_tokenize(jit(f))` once every time you cross a JIT boundary and not worry about tokens otherwise.

Reply:
Good point. Let me just remove it.

Reply:
Oh, actually I use it to rewrite the `cond` function in the `jax.lax.cond` transform. `cond` can only return a bool and not a token, so we set the token in `auto_tokenize` so that we at least get consistent ordering entering and within the method.